File: app/src/main/java/com/bongorian/signa1/MediaThumbnail.java
package com.bongorian.signa1;

import android.graphics.Bitmap;
import android.net.Uri;
import android.os.CancellationSignal;
import android.util.Size;
import android.view.*;
import android.widget.*;
import java.util.concurrent.*;

/** Loads only the app's last saved URI. No photo-library permission or main-thread decoding. */
final class MediaThumbnail extends FrameLayout {
    final MainActivity activity;final ImageView image;final TextView badge;
    final ExecutorService worker=Executors.newSingleThreadExecutor();
    CancellationSignal cancellation;Uri requested;boolean loading;int revision;boolean disposed;volatile boolean hasThumbnail;
    MediaThumbnail(MainActivity a){super(a);activity=a;setBackground(a.bg(MainActivity.PANEL,0xff394039));setClipToOutline(true);setClickable(true);
        image=new ImageView(a);image.setScaleType(ImageView.ScaleType.CENTER_CROP);addView(image,new LayoutParams(-1,-1));
        badge=a.text("",8,MainActivity.WHITE);badge.setGravity(Gravity.CENTER);badge.setBackground(a.detailBg(0xCC0A0C0D,0));LayoutParams bp=new LayoutParams(-2,a.dp(17),Gravity.BOTTOM|Gravity.END);bp.setMargins(0,0,a.dp(3),a.dp(3));badge.setPadding(a.dp(4),0,a.dp(4),0);addView(badge,bp);placeholder();
    }
    void placeholder(){hasThumbnail=false;image.setScaleType(ImageView.ScaleType.CENTER);image.setImageResource(R.drawable.ic_gallery);badge.setVisibility(GONE);setContentDescription(activity.getString(R.string.ui_open_saved_photos_and_videos));}
    void load(Uri uri,boolean video){if(disposed)return;if(uri!=null&&uri.equals(requested)&&(loading||hasThumbnail))return;requested=uri;loading=uri!=null;int ticket=++revision;if(cancellation!=null)cancellation.cancel();placeholder();if(uri==null)return;
        setContentDescription(video?activity.getString(R.string.ui_open_the_last_saved_video):activity.getString(R.string.ui_open_the_last_saved_photo));setTooltipText(getContentDescription());
        CancellationSignal signal=new CancellationSignal();cancellation=signal;
        worker.execute(()->{Bitmap bitmap=null;String type=null;try{type=activity.getContentResolver().getType(uri);bitmap=activity.getContentResolver().loadThumbnail(uri,new Size(192,192),signal);}catch(Exception ignored){}
            Bitmap result=bitmap;boolean sequence="application/zip".equals(type);boolean raw=sequence||(type!=null&&type.contains("dng"));activity.handler.post(()->{if(disposed||ticket!=revision){if(result!=null)result.recycle();return;}
                loading=false;if(result==null)placeholder();if(result!=null){hasThumbnail=true;image.setScaleType(ImageView.ScaleType.CENTER_CROP);image.setImageBitmap(result);}
                badge.setText(sequence?"RAW ZIP":video?"▶":raw?"RAW":"");if(sequence)setContentDescription(activity.getString(R.string.ui_open_the_last_saved_raw_video_zip));badge.setVisibility(video||raw?VISIBLE:GONE);
            });
        });
    }
    void dispose(){disposed=true;revision++;if(cancellation!=null)cancellation.cancel();worker.shutdownNow();image.setImageDrawable(null);}
}

File: app/src/main/kotlin/com/bongorian/signa1/MediaThumbnail.kt
package com.bongorian.signa1

import android.graphics.Bitmap
import android.net.Uri
import android.os.CancellationSignal
import android.util.Size
import android.view.Gravity
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.TextView
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors

/** Loads only the app's last saved URI. No photo-library permission or main-thread decoding. */
internal class MediaThumbnail(val activity: MainActivity) : FrameLayout(activity) {
    val image: ImageView
    val badge: TextView
    val worker: ExecutorService = Executors.newSingleThreadExecutor()
    var cancellation: CancellationSignal? = null
    var requested: Uri? = null
    var loading: Boolean = false
    var revision: Int = 0
    var disposed: Boolean = false

    @Volatile var hasThumbnail: Boolean = false

    init {
        setBackground(activity.bg(MainActivity.PANEL, -0xc6bfc7))
        setClipToOutline(true)
        setClickable(true)
        image = ImageView(activity)
        image.setScaleType(ImageView.ScaleType.CENTER_CROP)
        addView(image, LayoutParams(-1, -1))
        badge = activity.text("", 8, MainActivity.WHITE)
        badge.setGravity(Gravity.CENTER)
        badge.setBackground(activity.detailBg(-0x33f5f3f3, 0))
        val bp = LayoutParams(-2, activity.dp(17f), Gravity.BOTTOM or Gravity.END)
        bp.setMargins(0, 0, activity.dp(3f), activity.dp(3f))
        badge.setPadding(activity.dp(4f), 0, activity.dp(4f), 0)
        addView(badge, bp)
        placeholder()
    }

    fun placeholder() {
        hasThumbnail = false
        image.setScaleType(ImageView.ScaleType.CENTER)
        image.setImageResource(R.drawable.ic_gallery)
        badge.setVisibility(GONE)
        setContentDescription(activity.getString(R.string.ui_open_saved_photos_and_videos))
    }

    fun load(uri: Uri?, video: Boolean) {
        if (disposed) return
        if (uri != null && uri == requested && (loading || hasThumbnail)) return
        requested = uri
        loading = uri != null
        val ticket = ++revision
        if (cancellation != null) cancellation!!.cancel()
        placeholder()
        if (uri == null) return
        setContentDescription(
            if (video) activity.getString(R.string.ui_open_the_last_saved_video)
            else activity.getString(R.string.ui_open_the_last_saved_photo)
        )
        setTooltipText(getContentDescription())
        val signal = CancellationSignal()
        cancellation = signal
        worker.execute(
            Runnable@{
                var bitmap: Bitmap? = null
                var type: String? = null
                try {
                    type = activity.contentResolver.getType(uri)
                    bitmap = activity.contentResolver.loadThumbnail(uri, Size(192, 192), signal)
                } catch (ignored: Exception) {}
                val result = bitmap
                val sequence = "application/zip" == type
                val raw = sequence || (type != null && type.contains("dng"))
                activity.handler.post(
                    Runnable@{
                        if (disposed || ticket != revision) {
                            if (result != null) result.recycle()
                            return@Runnable
                        }
                        loading = false
                        if (result == null) placeholder()
                        if (result != null) {
                            hasThumbnail = true
                            image.setScaleType(ImageView.ScaleType.CENTER_CROP)
                            image.setImageBitmap(result)
                        }
                        badge.setText(
                            if (sequence) "RAW ZIP" else if (video) "▶" else if (raw) "RAW" else ""
                        )
                        if (sequence)
                            setContentDescription(
                                activity.getString(R.string.ui_open_the_last_saved_raw_video_zip)
                            )
                        badge.setVisibility(if (video || raw) VISIBLE else GONE)
                    }
                )
            }
        )
    }

    fun dispose() {
        disposed = true
        revision++
        if (cancellation != null) cancellation!!.cancel()
        worker.shutdownNow()
        image.setImageDrawable(null)
    }
}
